# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [48]:
import pandas as pd # trabajar con dataframes
import matplotlib.pyplot as plt # gráficos estándar
import seaborn as sns # librería para gráficos avanzados basado en matplotlib
import numpy as np # librería/biblioteca para elementos matemáticos. 
from datetime import datetime
from sklearn.model_selection import train_test_split

%matplotlib inline

pd.set_option('display.max_columns', None)



# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [2]:
data = pd.read_csv("austin_weather.csv")
df = data.copy()

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
df.shape

(1319, 21)

In [4]:
df.columns


Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'PrecipitationSumInches', 'Events'],
      dtype='object')

In [5]:
df.describe().round(2)


TempHighF  TempAvgF  TempLowF
count    1319.00   1319.00   1319.00
mean       80.86     70.64     59.90
std        14.77     14.05     14.19
min        32.00     29.00     19.00
25%        72.00     62.00     49.00
50%        83.00     73.00     63.00
75%        92.00     83.00     73.00
max       107.00     93.00     81.00

In [6]:
df.describe(exclude='number')

Date DewPointHighF DewPointAvgF DewPointLowF  \
count         1319          1319         1319         1319   
unique        1319            64           66           73   
top     2013-12-21            74           71           66   
freq             1           112           84           62   

       HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
count                 1319               1319               1319   
unique                  58                 69                 82   
top                    100                 64                 52   
freq                   221                 51                 46   

       SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
count                        1319                      1319   
unique                        105                       101   
top                         30.07                     29.95   
freq                           46                        49   

       SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
count                       1319                1319               1319   
unique                       105                   5                 10   
top                        29.88                  10                 10   
freq                          48                1302                859   

       VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
count                1319        1319       1319        1319   
unique                 12          22         13          37   
top                    10          13          4          20   
freq                  611         212        239         122   

       PrecipitationSumInches Events  
count                    1319   1319  
unique                    114      9  
top                         0         
freq                      870    903

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Date                        1319 non-null   object
 1   TempHighF                   1319 non-null   int64 
 2   TempAvgF                    1319 non-null   int64 
 3   TempLowF                    1319 non-null   int64 
 4   DewPointHighF               1319 non-null   object
 5   DewPointAvgF                1319 non-null   object
 6   DewPointLowF                1319 non-null   object
 7   HumidityHighPercent         1319 non-null   object
 8   HumidityAvgPercent          1319 non-null   object
 9   HumidityLowPercent          1319 non-null   object
 10  SeaLevelPressureHighInches  1319 non-null   object
 11  SeaLevelPressureAvgInches   1319 non-null   object
 12  SeaLevelPressureLowInches   1319 non-null   object
 13  VisibilityHighMiles         1319 non-null   obje

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [8]:
df["DewPointAvgF"].unique()


array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [9]:
# Your observation here


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [10]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [11]:
def clean_and_transform():
    for column in wrong_type_columns[1:]:
        df[column] = df[column].apply(lambda x: x.replace('-', '0') if '-' in str(x) else x)
        df[column] = pd.to_numeric(df[column], errors='coerce')
    return df.info()

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [12]:
clean_and_transform()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        1319 non-null   object 
 1   TempHighF                   1319 non-null   int64  
 2   TempAvgF                    1319 non-null   int64  
 3   TempLowF                    1319 non-null   int64  
 4   DewPointHighF               1319 non-null   object 
 5   DewPointAvgF                1319 non-null   int64  
 6   DewPointLowF                1319 non-null   int64  
 7   HumidityHighPercent         1319 non-null   int64  
 8   HumidityAvgPercent          1319 non-null   int64  
 9   HumidityLowPercent          1319 non-null   int64  
 10  SeaLevelPressureHighInches  1319 non-null   float64
 11  SeaLevelPressureAvgInches   1319 non-null   float64
 12  SeaLevelPressureLowInches   1319 non-null   float64
 13  VisibilityHighMiles         1319 

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [13]:
missing_data_rows = df.isnull().any(axis=1)

In [14]:
missing_values = df[missing_data_rows]
missing_values

Date  TempHighF  TempAvgF  TempLowF DewPointHighF  DewPointAvgF  \
4     2013-12-25         58        50        41            44            40   
6     2013-12-27         60        53        45            41            39   
7     2013-12-28         62        51        40            43            39   
42    2014-02-01         76        66        55            62            59   
51    2014-02-10         60        48        35            49            36   
...          ...        ...       ...       ...           ...           ...   
1269  2017-06-12         94        85        75            72            68   
1288  2017-07-01         98        89        79            76            72   
1304  2017-07-17         98        88        77            75            71   
1308  2017-07-21        104        91        77            74            68   
1313  2017-07-26        103        91        78            75            70   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
4               36                   86                  71   
6               37                   83                  65   
7               33                   92                  64   
42              41                   81                  71   
51              30                   82                  74   
...            ...                  ...                 ...   
1269            62                   87                  62   
1288            66                   85                  62   
1304            66                   88                  63   
1308            61                   85                  56   
1313            61                   91                  58   

      HumidityLowPercent  SeaLevelPressureHighInches  \
4                     56                       30.41   
6                     47                       30.46   
7                     36                       30.33   
42                    60                       29.91   
51                    66                       30.23   
...                  ...                         ...   
1269                  36                       29.96   
1288                  38                       30.13   
1304                  38                       29.94   
1308                  27                       30.01   
1313                  25                       30.13   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
4                         30.33                      30.27   
6                         30.39                      30.34   
7                         30.17                      30.04   
42                        29.81                      29.75   
51                        30.15                      30.02   
...                         ...                        ...   
1269                      29.89                      29.81   
1288                      30.02                      29.92   
1304                      29.90                      29.84   
1308                      29.95                      29.86   
1313                      30.03                      29.93   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
4                      10                  10                   7   
6                      10                   9                   7   
7                      10                  10                   7   
42                     10                  10                   9   
51                     10                   8                   4   
...                   ...                 ...                 ...   
1269                   10                  10                   2   
1288                   10                   9                   6   
1304                   10                  10                  10   
1308                   10                  10                  10   
1313                   10                  10                  10   

      WindHighMPH  WindAvgMPH  WindGustMPH  PrecipitationSu

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [15]:
rows_austin,columns_austin = df.shape
rows_missingv,columns_missingv = missing_values.shape
print(f'There is {rows_austin} rows in Austin and {rows_missingv} in missing_values') 

There is 1319 rows in Austin and 124 in missing_values


#### Calculate the ratio of missing rows to total rows

In [16]:
ratio = rows_missingv/rows_austin
ratio

0.09401061410159212

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [17]:
df.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   0
DewPointAvgF                    0
DewPointLowF                    0
HumidityHighPercent             0
HumidityAvgPercent              0
HumidityLowPercent              0
SeaLevelPressureHighInches      0
SeaLevelPressureAvgInches       0
SeaLevelPressureLowInches       0
VisibilityHighMiles             0
VisibilityAvgMiles              0
VisibilityLowMiles              0
WindHighMPH                     0
WindAvgMPH                      0
WindGustMPH                     0
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [18]:
df2 = df.drop("PrecipitationSumInches", axis=1)
df2.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [19]:
df2.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45            67            49   
1  2013-12-22         56        48        39            43            36   
2  2013-12-23         58        45        32            31            27   
3  2013-12-24         61        46        31            36            28   
4  2013-12-25         58        50        41            44            40   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0            43                   93                  75                  57   
1            28                   93                  68                  43   
2            23                   76                  52                  27   
3            21                   89                  56                  22   
4            36                   86                  71                  56   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                   10                   7   
1                      29.87                   10                  10   
2                      30.41                   10                  10   
3                      30.30                   10                  10   
4                      30.27                   10                  10   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                   2           20           4           31   
1                   5           16           6           25   
2                  10            8           3           12   
3                   7           12           4           20   
4                   7           10           2           16   

                Events  
0  Rain , Thunderstorm  
1                       
2                       
3                       
4

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [20]:
df_fixed = df.interpolate()

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [21]:
df_fixed.isna().sum()


Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
PrecipitationSumInches        0
Events                        0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [22]:
df["Events"].value_counts()


Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [23]:
print("The largest number of events were 3: Fog , Rain , Thunderstorm")

The largest number of events were 3: Fog , Rain , Thunderstorm


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [24]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

for event in event_list:
    df_fixed[event] = 0

df_fixed

Date  TempHighF  TempAvgF  TempLowF DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45            67            49   
1     2013-12-22         56        48        39            43            36   
2     2013-12-23         58        45        32            31            27   
3     2013-12-24         61        46        31            36            28   
4     2013-12-25         58        50        41            44            40   
...          ...        ...       ...       ...           ...           ...   
1314  2017-07-27        103        89        75            71            67   
1315  2017-07-28        105        91        76            71            64   
1316  2017-07-29        107        92        77            72            64   
1317  2017-07-30        106        93        79            70            68   
1318  2017-07-31         99        88        77            66            61   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0               43                   93                  75   
1               28                   93                  68   
2               23                   76                  52   
3               21                   89                  56   
4               36                   86                  71   
...            ...                  ...                 ...   
1314            61                   82                  54   
1315            55                   87                  54   
1316            55                   82                  51   
1317            63                   69                  48   
1318            54                   64                  43   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                     57                       29.86   
1                     43                       30.41   
2                     27                       30.56   
3                     22                       30.56   
4                     56                       30.41   
...                  ...                         ...   
1314                  25                       30.04   
1315                  20                       29.97   
1316                  19                       29.91   
1317                  27                       29.96   
1318                  22                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                      10                   7                   2   
1                      10                  10                   5   
2                      10                  10                  10   
3                      10                  10                   7   
4                      10                  10                   7   
...                   ...                 ...                 ...   
1314                   10                  10                  10   
1315                   10                  10                  10   
1316                   10                  10                  10   
1317                   10                  10                  10   
1318                   10                  10                  10   

      WindHighMPH  WindAvgMPH  WindGustMPH  PrecipitationSu

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [25]:
for event in event_list:
    df_fixed[event] = df_fixed['Events'].str.contains(event).astype(int)


#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [26]:
df_fixed.sample(15)


Date  TempHighF  TempAvgF  TempLowF DewPointHighF  DewPointAvgF  \
520   2015-05-25         82        74        66            74            69   
811   2016-03-11         71        65        59            63            61   
91    2014-03-22         77        69        61            65            62   
677   2015-10-29         81        66        51            64            54   
398   2015-01-23         51        45        38            42            38   
820   2016-03-20         64        53        42            32            28   
698   2015-11-19         79        63        46            51            46   
630   2015-09-12         91        80        69            72            57   
803   2016-03-03         83        73        62            64            50   
1138  2017-02-01         82        67        51            59            52   
161   2014-05-31         93        82        70            69            67   
864   2016-05-03         78        67        55            51            48   
1107  2017-01-01         73        63        53            60            52   
866   2016-05-05         89        74        59            56            53   
424   2015-02-18         66        49        32            34            29   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
520             66                  100                  87   
811             57                  100                  87   
91              51                  100                  80   
677             49                   96                  66   
398             35                   93                  77   
820             25                   62                  44   
698             41                   86                  59   
630             46                   87                  55   
803             33                   87                  53   
1138            45                  100                  68   
161             65                   84                  63   
864             46                   86                  60   
1107            39                   97                  74   
866             50                   72                  51   
424             25                   89                  56   

      HumidityLowPercent  SeaLevelPressureHighInches  \
520                   74                       29.92   
811                   73                       29.97   
91                    60                       30.02   
677                   35                       29.97   
398                   61                       30.32   
820                   25                       30.47   
698                   32                       30.15   
630                   23                       30.08   
803                   18                       30.13   
1138                  35                       30.09   
161                   41                       29.89   
864                   33                       30.14   
1107                  43                       29.93   
866                   29                       30.07   
424                   22                       30.39   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
520                       29.84                      29.75   
811                       29.93                      29.87   
91                        29.95                      29.88   
677                       29.89                      29.83   
398                       30.26                      30.20   
820                       30.39                      30.30   
698                       30.08                      29.97   
630                       30.02                      29.96   
803                       29.97                      29.83   
1138                      29.98                      29.89   
161                       29.85                      29.82   
864                       30.08                      30.00   
1107                      29.85    

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [28]:
df_fixed.drop("Events", axis=1, inplace = True)


In [29]:
df_fixed

Date  TempHighF  TempAvgF  TempLowF DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45            67            49   
1     2013-12-22         56        48        39            43            36   
2     2013-12-23         58        45        32            31            27   
3     2013-12-24         61        46        31            36            28   
4     2013-12-25         58        50        41            44            40   
...          ...        ...       ...       ...           ...           ...   
1314  2017-07-27        103        89        75            71            67   
1315  2017-07-28        105        91        76            71            64   
1316  2017-07-29        107        92        77            72            64   
1317  2017-07-30        106        93        79            70            68   
1318  2017-07-31         99        88        77            66            61   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0               43                   93                  75   
1               28                   93                  68   
2               23                   76                  52   
3               21                   89                  56   
4               36                   86                  71   
...            ...                  ...                 ...   
1314            61                   82                  54   
1315            55                   87                  54   
1316            55                   82                  51   
1317            63                   69                  48   
1318            54                   64                  43   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                     57                       29.86   
1                     43                       30.41   
2                     27                       30.56   
3                     22                       30.56   
4                     56                       30.41   
...                  ...                         ...   
1314                  25                       30.04   
1315                  20                       29.97   
1316                  19                       29.91   
1317                  27                       29.96   
1318                  22                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                      10                   7                   2   
1                      10                  10                   5   
2                      10                  10                  10   
3                      10                  10                   7   
4                      10                  10                   7   
...                   ...                 ...                 ...   
1314                   10                  10                  10   
1315                   10                  10                  10   
1316                   10                  10                  10   
1317                   10                  10                  10   
1318                   10                  10                  10   

      WindHighMPH  WindAvgMPH  WindGustMPH  PrecipitationSu

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [44]:
df_fixed['Numeric_Date'] = df_fixed['Date'].apply(lambda x: datetime.strptime(x, '%Y-%m-%d').toordinal())

In [45]:
df_fixed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        1319 non-null   object 
 1   TempHighF                   1319 non-null   int64  
 2   TempAvgF                    1319 non-null   int64  
 3   TempLowF                    1319 non-null   int64  
 4   DewPointHighF               1319 non-null   object 
 5   DewPointAvgF                1319 non-null   int64  
 6   DewPointLowF                1319 non-null   int64  
 7   HumidityHighPercent         1319 non-null   int64  
 8   HumidityAvgPercent          1319 non-null   int64  
 9   HumidityLowPercent          1319 non-null   int64  
 10  SeaLevelPressureHighInches  1319 non-null   float64
 11  SeaLevelPressureAvgInches   1319 non-null   float64
 12  SeaLevelPressureLowInches   1319 non-null   float64
 13  VisibilityHighMiles         1319 

#### Print `austin_fixed` to check your `Date` column.

In [47]:
df_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45            67            49   
1  2013-12-22         56        48        39            43            36   
2  2013-12-23         58        45        32            31            27   
3  2013-12-24         61        46        31            36            28   
4  2013-12-25         58        50        41            44            40   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0            43                   93                  75                  57   
1            28                   93                  68                  43   
2            23                   76                  52                  27   
3            21                   89                  56                  22   
4            36                   86                  71                  56   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                   10                   7   
1                      29.87                   10                  10   
2                      30.41                   10                  10   
3                      30.30                   10                  10   
4                      30.27                   10                  10   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                   2           20           4           31   
1                   5           16           6           25   
2                  10            8           3           12   
3                   7           12           4           20   
4                   7           10           2           16   

   PrecipitationSumInches  Snow  Fog  Rain  Thunderstorm  Numeric_Date  
0                    0.46     0    0     1             1        735223  
1                    0.00     0    0     0             0        735224  
2                    0.00     0    0     0             0        735225  
3                    0.00     0    0     0             0        735226  
4                    0.00     0    0     0             0        735227

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [49]:
X = df_fixed.drop(columns=["TempAvgF"], axis=1)
y = df_fixed["TempAvgF"]

In the next cell, import `train_test_split` from `sklearn.model_selection`

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [51]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [52]:
print(f'100% of our data: {len(df_fixed)}.')
print(f'80% for training data: {len(X_train)}.')
print(f'20% for test data: {len(X_test)}.')


100% of our data: 1319.
80% for training data: 1055.
20% for test data: 264.


In [53]:
ts_rows = len(X_test)

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
